File: Segundo/PDOO/CivitasRuby/Civitas/JugadorEspeculador.rb
# encodign:utf-8

require_relative "Jugador"

module Civitas

    class JugadorEspeculador < Jugador

        attr_accessor :FactorEspeculador

        def initialize (f)
            @fianza = f
            @FactorEspeculador = 2
        end

        def self.nuevoEspeculador(otro, f)
            nuevoJugador = self.new(f)
            nuevoJugador.copiaDe(otro)
            nuevoJugador.hotelesMax = nuevoJugador.hotelesMax * nuevoJugador.FactorEspeculador
            nuevoJugador.casasMax = nuevoJugador.casasMax * nuevoJugador.FactorEspeculador
            nuevoJugador.propiedades.each do |propiedad|
                propiedad.actualizarPropietario(nuevoJugador)
            end
            
            return nuevoJugador
        end

        def quieroPagar
            ok = false
            msg1 = "Desea pagar la fianza de " + @fianza.to_s + "? (1 -> si, 0 -> no)"
            msg2 = "Valor incorrecto"
            begin
              print msg1
              cadena = gets.chomp
              begin
                if (cadena =~ /\A\d+\Z/)
                  numero = cadena.to_i
                  ok = true
                else
                  raise IOError
                end
              rescue IOError
                puts msg2
              end
              if (ok)
                if (numero > 1)
                  ok = false
                end
              end
            end while (!ok)
            
            if numero == 1
                return true
            else
                return false
            end
        end

        def debeSerEncarcelado
            if @encarcelado
                return false
            elsif tieneSalvoconducto()
                perderSalvoconducto()
                Diario.instance.ocurre_evento("El jugador " + @nombre + " se libra de la cárcel a cambio del salvoconducto.")
                return false
            elsif puedoGastar(@fianza)
                if quieroPagar()
                    paga(@fianza)
                    Diario.instance.ocurre_evento("El jugador " + @nombre + " se libra de la cárcel pagando la fianza de " + @fianza.to_s + ".")
                    return false
                else
                    return true
                end
            else
                return true
            end
        end

        def pagaImpuesto(cantidad)
            if @encarcelado
                return false
            else
                return paga(cantidad/@FactorEspeculador)
            end
        end

        def toString
            return "\nEl jugador especulador actual es " + @nombre + " y tiene un saldo de " + @saldo.to_s + ".\n"
        end

    end

end

File: Segundo/PDOO/CivitasJava/src/main/java/civitas/JugadorEspeculador.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

import java.util.Scanner;


/**
 *
 * @author jmramirez
 */
public class JugadorEspeculador extends Jugador {
    private final int FactorEspeculador = 2;
    private final float fianza;
    
    JugadorEspeculador(Jugador jugador, float f) {
        super(jugador);
        fianza = f;
        HotelesMax = HotelesMax*FactorEspeculador;
        CasasMax = CasasMax*FactorEspeculador;
        actualizaPropietarioPorConversion();
    }
    
    Boolean pagaImpuesto(float cantidad) {
        if (isEncarcelado()) {
            return false;
        } else {
            return paga(cantidad/FactorEspeculador);
        }
    }
    
    Boolean debeSerEncarcelado() {
        if (isEncarcelado()) {
            return false;
        } else if (tieneSalvoconducto()) {
            perderSalvoconducto();
            Diario.getInstance().ocurreEvento("El jugador especulador " + getNombre() + " se libra de la cárcel a cambio del salvoconducto");
            return false;
        } else if (puedoGastar(fianza)){
            System.out.println("Desea Pagar una fianza de " + fianza + " y así evitar la cárcel? (y/n)");
            Scanner in = new Scanner(System.in);
            String cadena = in.nextLine();
            if ("y".equals(cadena)) {
                Diario.getInstance().ocurreEvento("El jugador especulador " + getNombre() + " se libra de la cárcel pagando una fianza");
                paga(fianza);
                return false;
            } else {
                return true;
            }
        } else {
            return true;
        }
    }
    
    public String toString() {
        return "el jugador especulador " + getNombre() + " posee " + getPropiedades().size() + " propiedades y tiene un saldo de " + getSaldo() + ".";
    }
}
